File: app/src/main/java/com/manimarank/spell4wiki/activities/SettingsActivity.java
package com.manimarank.spell4wiki.activities;

import android.app.AlertDialog;
import android.os.Bundle;
import android.text.Html;
import android.text.TextUtils;
import android.text.method.LinkMovementMethod;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

import com.manimarank.spell4wiki.R;
import com.manimarank.spell4wiki.utils.constants.EnumTypeDef.ListMode;
import com.manimarank.spell4wiki.databases.DBHelper;
import com.manimarank.spell4wiki.databases.dao.WikiLangDao;
import com.manimarank.spell4wiki.databases.entities.WikiLang;
import com.manimarank.spell4wiki.fragments.LanguageSelectionFragment;
import com.manimarank.spell4wiki.listerners.OnLanguageSelectionListener;
import com.manimarank.spell4wiki.utils.PrefManager;
import com.manimarank.spell4wiki.utils.WikiLicense;

import java.util.Arrays;


public class SettingsActivity extends AppCompatActivity {

    private PrefManager pref;
    private WikiLangDao wikiLangDao = null;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_settings);
        setTitle(getString(R.string.settings));

        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        TextView txtTitleLicense = findViewById(R.id.txtTitleLicense);
        TextView txtSpell4WikiLang = findViewById(R.id.txtSpell4WikiLang);
        TextView txtSpell4WordListLang = findViewById(R.id.txtSpell4WordListLang);
        TextView txtSpell4WordLang = findViewById(R.id.txtSpell4WordLang);
        TextView txtWiktionaryLang = findViewById(R.id.txtWiktionaryLang);
        TextView txtLicenseOfUploadAudio = findViewById(R.id.txtLicenseOfUploadAudio);
        TextView txtLicenseOfUploadAudioLegalCode = findViewById(R.id.txtLicenseOfUploadAudioLegalCode);


        View layoutSpell4WikiLang = findViewById(R.id.layoutSpell4WikiLang);
        View layoutSpell4WordListLang = findViewById(R.id.layoutSpell4WordListLang);
        View layoutSpell4WordLang = findViewById(R.id.layoutSpell4WordLang);
        View layoutWiktionaryLang = findViewById(R.id.layoutWiktionaryLang);
        View layoutLicenseOfUploadAudio = findViewById(R.id.layoutLicenseOfUploadAudio);

        pref = new PrefManager(getApplicationContext());
        wikiLangDao = DBHelper.getInstance(getApplicationContext()).getAppDatabase().getWikiLangDao();
        if (pref.getIsAnonymous()) {
            txtTitleLicense.setVisibility(View.GONE);
            layoutSpell4WikiLang.setVisibility(View.GONE);
            layoutSpell4WordListLang.setVisibility(View.GONE);
            layoutSpell4WordLang.setVisibility(View.GONE);
            layoutLicenseOfUploadAudio.setVisibility(View.GONE);
        }

        updateLanguageView(txtSpell4WikiLang, pref.getLanguageCodeSpell4Wiki());
        updateLanguageView(txtSpell4WordListLang, pref.getLanguageCodeSpell4WordList());
        updateLanguageView(txtSpell4WordLang, pref.getLanguageCodeSpell4Word());
        updateLanguageView(txtWiktionaryLang, pref.getLanguageCodeWiktionary());

        layoutSpell4WikiLang.setOnClickListener(v -> {
            OnLanguageSelectionListener callback = langCode -> {
                txtSpell4WikiLang.setText(pref.getLanguageCodeSpell4Wiki());
            };
            LanguageSelectionFragment languageSelectionFragment = new LanguageSelectionFragment(this);
            languageSelectionFragment.init(callback, ListMode.SPELL_4_WIKI);
            languageSelectionFragment.show(getSupportFragmentManager(), languageSelectionFragment.getTag());
        });

        layoutSpell4WordListLang.setOnClickListener(v -> {
            OnLanguageSelectionListener callback = langCode -> {
                updateLanguageView(txtSpell4WordListLang, pref.getLanguageCodeSpell4WordList());
            };
            LanguageSelectionFragment languageSelectionFragment = new LanguageSelectionFragment(this);
            languageSelectionFragment.init(callback, ListMode.SPELL_4_WORD_LIST);
            languageSelectionFragment.show(getSupportFragmentManager(), languageSelectionFragment.getTag());
        });


        layoutSpell4WordLang.setOnClickListener(v -> {
            OnLanguageSelectionListener callback = langCode -> {
                updateLanguageView(txtSpell4WordLang, pref.getLanguageCodeSpell4Word());
            };
            LanguageSelectionFragment languageSelectionFragment = new LanguageSelectionFragment(this);
            languageSelectionFragment.init(callback, ListMode.SPELL_4_WORD);
            languageSelectionFragment.show(getSupportFragmentManager(), languageSelectionFragment.getTag());
        });

        layoutWiktionaryLang.setOnClickListener(v -> {
            OnLanguageSelectionListener callback = langCode -> {
                updateLanguageView(txtWiktionaryLang, pref.getLanguageCodeWiktionary());
            };
            LanguageSelectionFragment languageSelectionFragment = new LanguageSelectionFragment(this);
            languageSelectionFragment.init(callback, ListMode.WIKTIONARY);
            languageSelectionFragment.show(getSupportFragmentManager(), languageSelectionFragment.getTag());
        });

        updateLicenseView(txtLicenseOfUploadAudio, txtLicenseOfUploadAudioLegalCode);
        layoutLicenseOfUploadAudio.setOnClickListener(v -> {
            // setup the alert builder
            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setTitle(R.string.license_choose_alert);// add a radio button list

            String[] licensePrefList = {
                    WikiLicense.LicensePrefs.CC_0,
                    WikiLicense.LicensePrefs.CC_BY_3,
                    WikiLicense.LicensePrefs.CC_BY_SA_3,
                    WikiLicense.LicensePrefs.CC_BY_4,
                    WikiLicense.LicensePrefs.CC_BY_SA_4
            };

            String[] licenseList = {
                    getString(R.string.license_name_cc_zero),
                    getString(R.string.license_name_cc_by_three),
                    getString(R.string.license_name_cc_by_sa_three),
                    getString(R.string.license_name_cc_by_four),
                    getString(R.string.license_name_cc_by_sa_four)
            };

            int checkedItem = Arrays.asList(licensePrefList).indexOf(pref.getUploadAudioLicense());
            builder.setSingleChoiceItems(licenseList, checkedItem, (dialog, which) -> {
                pref.setUploadAudioLicense(licensePrefList[which]);
                updateLicenseView(txtLicenseOfUploadAudio, txtLicenseOfUploadAudioLegalCode);
                dialog.dismiss();
            });
            builder.setNegativeButton(getString(R.string.cancel), null);
            AlertDialog dialog = builder.create();
            dialog.show();
        });

    }

    private void updateLanguageView(TextView txtView, String languageCode) {
        if (languageCode != null && txtView != null && wikiLangDao != null) {
            WikiLang wikiLang = wikiLangDao.getWikiLanguageWithCode(languageCode);
            String value = "";
            if (wikiLang != null && !TextUtils.isEmpty(wikiLang.getName()))
                value = wikiLang.getLocalName() + " - " + wikiLang.getName() + "(" + languageCode + ")";
            txtView.setText(value);
        }
    }

    private void updateLicenseView(TextView txtLicenseOfUploadAudio, TextView txtLicenseOfUploadAudioLegalCode) {
        txtLicenseOfUploadAudio.setText(getString(WikiLicense.licenseNameId(pref.getUploadAudioLicense())));
        txtLicenseOfUploadAudioLegalCode.setMovementMethod(LinkMovementMethod.getInstance());
        String ccLegalInfo = "(<a href=" + WikiLicense.licenseUrlFor(pref.getUploadAudioLicense()) + "><font color='" + ContextCompat.getColor(getApplicationContext(), R.color.w_green) + "'>legal code</font></a>)";
        txtLicenseOfUploadAudioLegalCode.setText(Html.fromHtml(ccLegalInfo));
    }


    @Override
    public boolean onOptionsItemSelected(MenuItem menuItem) {
        if (menuItem.getItemId() == android.R.id.home) {
            finish();
            return true;
        }
        return (super.onOptionsItemSelected(menuItem));
    }

}

File: app/src/main/java/com/manimarank/spell4wiki/ui/settings/SettingsActivity.kt
package com.manimarank.spell4wiki.ui.settings

import android.app.AlertDialog
import android.content.DialogInterface
import android.os.Bundle
import android.text.TextUtils
import android.text.method.LinkMovementMethod
import android.view.MenuItem
import android.widget.SeekBar
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.core.text.HtmlCompat
import com.manimarank.spell4wiki.R
import com.manimarank.spell4wiki.data.db.DBHelper
import com.manimarank.spell4wiki.data.db.dao.WikiLangDao
import com.manimarank.spell4wiki.data.prefs.PrefManager
import com.manimarank.spell4wiki.ui.common.BaseActivity
import com.manimarank.spell4wiki.ui.dialogs.AppLanguageDialog.getSelectedLanguage
import com.manimarank.spell4wiki.ui.dialogs.AppLanguageDialog.show
import com.manimarank.spell4wiki.ui.languageselector.LanguageSelectionFragment
import com.manimarank.spell4wiki.ui.listerners.OnLanguageSelectionListener
import com.manimarank.spell4wiki.ui.recordaudio.RecordAudioActivity
import com.manimarank.spell4wiki.utils.EdgeToEdgeUtils.setupStatusBarHandling
import com.manimarank.spell4wiki.utils.WikiLicense
import com.manimarank.spell4wiki.utils.WikiLicense.licenseNameId
import com.manimarank.spell4wiki.utils.WikiLicense.licenseUrlFor
import com.manimarank.spell4wiki.utils.constants.AppConstants
import com.manimarank.spell4wiki.utils.constants.ListMode
import com.manimarank.spell4wiki.utils.extensions.showLicenseChooseDialog
import com.manimarank.spell4wiki.utils.makeGone
import com.manimarank.spell4wiki.databinding.ActivitySettingsBinding
import java.util.*

class SettingsActivity : BaseActivity() {
    private lateinit var binding: ActivitySettingsBinding
    private lateinit var pref: PrefManager
    private var wikiLangDao: WikiLangDao? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivitySettingsBinding.inflate(layoutInflater)
        setContentView(binding.root)

        // Setup proper status bar handling
        setupStatusBarHandling(binding.root)

        title = getString(R.string.settings)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        pref = PrefManager(applicationContext)
        wikiLangDao = DBHelper.getInstance(applicationContext).appDatabase.wikiLangDao
        if (pref.isAnonymous == true) {
            binding.txtTitleLicense.makeGone()
            // binding.layoutSpell4WikiLang.makeGone()
            binding.layoutLicenseOfUploadAudio.makeGone()
            binding.layoutRunFilter.makeGone()
        }
        updateLanguageView(binding.txtSpell4WikiLang, pref.languageCodeSpell4WikiAll)
        binding.layoutSpell4WikiLang.setOnClickListener {
            val callback = object : OnLanguageSelectionListener {
                override fun onCallBackListener(langCode: String?) {
                    updateLanguageView(binding.txtSpell4WikiLang, pref.languageCodeSpell4WikiAll)
                }
            }
            val languageSelectionFragment = LanguageSelectionFragment(this)
            languageSelectionFragment.init(callback, ListMode.SPELL_4_WIKI_ALL)
            languageSelectionFragment.show(supportFragmentManager)
        }


        updateLicenseView(binding.txtLicenseOfUploadAudio, binding.txtLicenseOfUploadAudioLegalCode)



        binding.layoutLicenseOfUploadAudio.setOnClickListener {
            showLicenseChooseDialog {
                updateLicenseView(binding.txtLicenseOfUploadAudio, binding.txtLicenseOfUploadAudioLegalCode)
            } }

        binding.txtAppLanguage.text = getSelectedLanguage()
        binding.layoutLanguageOfApp.setOnClickListener { show(this@SettingsActivity) }

        // Theme selection setting
        updateThemeView()
        binding.layoutThemeSelection.setOnClickListener {
            showThemeSelectionDialog()
        }

        // Wiktionary cleanup setting
        binding.switchWiktionaryCleanup.isChecked = pref.isWiktionaryCleanupEnabled
        binding.switchWiktionaryCleanup.setOnCheckedChangeListener { _, isChecked ->
            pref.isWiktionaryCleanupEnabled = isChecked
        }

        binding.txtRfCount.text = getString(R.string.run_filter_settings_count, pref.runFilterNumberOfWordsToCheck
                ?: AppConstants.RUN_FILTER_NO_OF_WORDS_CHECK_COUNT)
        binding.seekBarRunFilterCount.progress = pref.runFilterNumberOfWordsToCheck
                ?: AppConstants.RUN_FILTER_NO_OF_WORDS_CHECK_COUNT
        binding.seekBarRunFilterCount.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            var finalProgress = pref.runFilterNumberOfWordsToCheck
                    ?: AppConstants.RUN_FILTER_NO_OF_WORDS_CHECK_COUNT

            override fun onProgressChanged(p0: SeekBar?, progress: Int, p2: Boolean) {
                finalProgress = if (progress <= 0) 1 else progress
                binding.txtRfCount.text = getString(R.string.run_filter_settings_count, finalProgress)
            }

            override fun onStartTrackingTouch(p0: SeekBar?) = Unit

            override fun onStopTrackingTouch(p0: SeekBar?) {
                pref.runFilterNumberOfWordsToCheck = finalProgress
            }

        })
    }

    private fun updateLanguageView(txtView: TextView, languageCode: String?) {
        if (languageCode != null) {
            val wikiLang = wikiLangDao?.getWikiLanguageWithCode(languageCode)
            var value = ""
            if (wikiLang != null && !TextUtils.isEmpty(wikiLang.name)) value = wikiLang.localName + " - " + wikiLang.name + " : " + languageCode
            txtView.text = value
        }
    }

    private fun updateThemeView() {
        val themeText = when (pref.themeMode) {
            PrefManager.ThemeMode.LIGHT -> getString(R.string.theme_light)
            PrefManager.ThemeMode.DARK -> getString(R.string.theme_dark)
            PrefManager.ThemeMode.SYSTEM -> getString(R.string.theme_system_default)
            else -> getString(R.string.theme_system_default)
        }
        binding.txtThemeSelection.text = themeText
    }

    private fun showThemeSelectionDialog() {
        val themes = arrayOf(
            getString(R.string.theme_light),
            getString(R.string.theme_dark),
            getString(R.string.theme_system_default)
        )

        val currentSelection = when (pref.themeMode) {
            PrefManager.ThemeMode.LIGHT -> 0
            PrefManager.ThemeMode.DARK -> 1
            PrefManager.ThemeMode.SYSTEM -> 2
            else -> 2
        }

        AlertDialog.Builder(this)
            .setTitle(getString(R.string.choose_theme))
            .setSingleChoiceItems(themes, currentSelection) { dialog, which ->
                val selectedTheme = when (which) {
                    0 -> PrefManager.ThemeMode.LIGHT
                    1 -> PrefManager.ThemeMode.DARK
                    2 -> PrefManager.ThemeMode.SYSTEM
                    else -> PrefManager.ThemeMode.SYSTEM
                }

                pref.themeMode = selectedTheme
                updateThemeView()

                // Recreate activity to apply theme change
                recreate()

                dialog.dismiss()
            }
            .setNegativeButton(getString(R.string.cancel), null)
            .show()
    }

    private fun updateLicenseView(txtLicenseOfUploadAudio: TextView, txtLicenseOfUploadAudioLegalCode: TextView) {
        txtLicenseOfUploadAudio.text = getString(licenseNameId(pref.uploadAudioLicense))
        txtLicenseOfUploadAudioLegalCode.movementMethod = LinkMovementMethod.getInstance()
        val ccLegalInfo = "(<a href=" + licenseUrlFor(pref.uploadAudioLicense) + "><font color='" + ContextCompat.getColor(applicationContext, R.color.w_green) + "'>legal code</font></a>)"
        txtLicenseOfUploadAudioLegalCode.text = HtmlCompat.fromHtml(ccLegalInfo, HtmlCompat.FROM_HTML_MODE_LEGACY)
    }

    override fun onOptionsItemSelected(menuItem: MenuItem): Boolean {
        if (menuItem.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(menuItem)
    }
}